modal/_partial_function.py

```diff
@@ -922,3 +922,174 @@ def wrapper(
         return pf
 
     return wrapper
+
+
+# NOTE: torchrun is currently exposed through modal.experimental, not the top-level namespace
+def _torchrun(
+    _warn_parentheses_missing=None,
+    *,
+    nproc_per_node: int = 1,
+    master_port: int = 29500,
+) -> Callable[
+    [Union[_PartialFunction[P, ReturnType, ReturnType], Callable[P, ReturnType]]],
+    _PartialFunction[P, ReturnType, ReturnType],
+]:
+    """Decorator that enables torchrun orchestration for distributed training.
+
+    When used with @modal.experimental.clustered, this decorator automatically handles
+    the torchrun orchestration, allowing you to write your training logic directly
+    in the decorated function without needing a separate script file.
+
+    The function will be called in two modes:
+    - Orchestration mode: Sets up and launches torchrun
+    - Data mode: Executes the actual training logic (called by torchrun)
+
+    Parameters:
+    nproc_per_node: int = 1
+        Number of processes to launch per node
+    master_port: int = 29500
+        Port for inter-process communication
+
+    Example:
+    ```python
+    @app.function(gpu="H100:8")
+    @modal.experimental.clustered(4, rdma=True)
+    @modal.experimental.torchrun(nproc_per_node=8)
+    def train():
+        # This code runs in data mode
+        import torch.distributed as dist
+        dist.init_process_group(backend="nccl")
+
+        # Your training logic here
+        ...
+    ```
+    """
+    if _warn_parentheses_missing is not None:
+        raise InvalidError(
+            "Positional arguments are not allowed. Did you forget parentheses? Suggestion: `@modal.experimental.torchrun()`."
+        )
+
+    def wrapper(
+        obj: Union[_PartialFunction[P, ReturnType, ReturnType], Callable[P, ReturnType]],
+    ) -> _PartialFunction[P, ReturnType, ReturnType]:
+        import functools
+        import os
+
+        # Get the original function
+        if isinstance(obj, _PartialFunction):
+            if obj.raw_f is None:
+                raise InvalidError("Cannot apply @modal.experimental.torchrun to a class-level decorator")
+            original_func = obj.raw_f
+        else:
+            original_func = obj
+
+        @functools.wraps(original_func)
+        def torchrun_wrapper(*args, **kwargs):
+            # Check if we're in data mode (torchrun has set these env vars)
+            if all(var in os.environ for var in ["RANK", "LOCAL_RANK", "WORLD_SIZE"]):
+                # We're in data mode - run the original function
+                return original_func(*args, **kwargs)
+            else:
+                # We're in orchestration mode
+                import sys
+                import subprocess
+                import tempfile
+                import json
+                import modal
+
+                # Get cluster info
+                cluster_info = modal.experimental.get_cluster_info()
```

Comment:
Once nice thing about this is we wouldn't necessarily need to expose this stuff in a public API (though we probably still would unless we think we'll cover all viable multi-node training frameworks with integrations I guess).